### PR TITLE
fix(compare): generic perfomance improve

### PR DESCRIPTION
**Pull request by Delagen**
# Purpose / Goal
Parser fix

# Type
Please mention the type of PR

* [x]Bug Fix
* [ ]Refactoring / Technology upgrade
* [ ]New Feature

**Note** : Please ensure that you've read contribution [guidelines](https://github.com/NaturalIntelligence/fast-xml-parser/blob/master/docs/CONTRIBUTING.md) before raising this PR. If your PR is in progress, please prepend `[WIP]` in PR title. Your PR will be reviewed when `[WIP]` will be removed from the PR title.

[Bookmark](https://github.com/NaturalIntelligence/fast-xml-parser/stargazers) this repository for further updates.

### CODE REVIEW

**Comment by Delagen** [comment]
@amitguptagwl Please merge and publish as can

**Comment by amitguptagwl** [comment]
Looks like I had forgotten to push some changes to GitHub after npm publish. It has caused code conflict in your PR.

**Comment by Delagen** [comment]
@amitguptagwl Please push it, I try to merge )

**Comment by Delagen** [comment]
Checked test: slice x 11,092,251 ops/sec ±3.81% (61 runs sampled)
Checked test: substr x 11,861,028 ops/sec ±0.96% (65 runs sampled)
Checked test: substring x 11,772,668 ops/sec ±1.47% (64 runs sampled)
Checked test: index x 5,805,558 ops/sec ±1.59% (64 runs sampled)

I think second faster

**Comment by Delagen** [comment]
https://www.measurethat.net/Benchmarks/Show/8327/0/slice-vs-substr-vs-substring-vs-index

**Comment by Delagen** [comment]
updated )

**Comment by amitguptagwl** [comment]
@Delagen As I said in the past as well, one thing in one PR. Can you please remove generated bundle, updated dependencies, etc.? We do it only at the time of publishing a version not for intermediate change. 

Just the code change, UTs, and doc change if applicable.

**Comment by Delagen** [comment]
Ok

**Comment by Delagen** [comment]
Removed last commit

**Comment by hata6502** [comment]
@Delagen 
Thank you so much!
I could parse XML correctly by your fork!

This XML file includes DOCTYPE
http://monoroch.net/kinshi/housouKinshiYougo.xml
```
<!DOCTYPE housouKinshiYougoList SYSTEM "http://monoroch.net/kinshi/housouKinshiYougo-1.0.dtd">
```

**Comment by amitguptagwl** [comment]
@Delagen can you please remove bundle files from this PR?

**Comment by amitguptagwl** [comment]
@Delagen I believe changes on line 216 in parser was sufficient to fix the change. Let me know if you think otherwise. Your 2nd test was repeated as I already implemented issue when I published 3.17.1. And first test was failing as result was missing with 'svg' tag. So I have published the changes in 3.17.2.

Thanks to highlight the issue and your effort. Let me know if you feel I missed something.

**Comment by Delagen** [comment]
substr comparison with at least 2 chars faster than separate indexed near 2 times, and they look more readable

**Comment by Delagen** [comment]
Moreover fix does not correct. Try this
```
const xmlData2 = "<?xml version='1.0' standalone='no'?><!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\" \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\" ><svg><metadata>[test]</metadata></svg>";
        const result2 = parser.parse(xmlData2, {
            ignoreAttributes: false,
            allowBooleanAttributes: true
        });
        console.log(result2);
```

**Comment by amitguptagwl** [comment]
Yes that is because I used `substr` instead of `substring` on line 215

https://github.com/NaturalIntelligence/fast-xml-parser/blob/master/src/xmlstr2xmlnode.js#L215

I didn't check the performance for both approaches. And I believe it'll not be significant. You can use anyone you feel more suitable.

File: spec/cdata_spec.js [hunk 1]
```diff
@@ -292,4 +292,33 @@ describe("XMLParser", function() {
         //console.log(JSON.stringify(result,null,4));
         expect(result).toEqual(expected);
     });
+
+    it("should parse XML with DOCTYPE without internal DTD", function() {
+        const xmlData = "<?xml version='1.0' standalone='no'?><!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\" \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\" ><svg><metadata>test</metadata></svg>";
+        const expected = {
```
Comment (amitguptagwl) [inline spec/cdata_spec.js:298]: Not expected output. This test must be failing. And this is already covered.

File: spec/cdata_spec.js [hunk 2]
```diff
@@ -292,4 +292,33 @@ describe("XMLParser", function() {
...
+        };
+
+        const result = parser.parse(xmlData, {
+            ignoreAttributes:       false,
+            allowBooleanAttributes: true
+        });
+        expect(result).toEqual(expected);
+    });
+
+    it("should parse XML with dashed elements, when - second symbol", function() {
```
Comment (amitguptagwl) [inline spec/cdata_spec.js:309]: This is already covered. You can just update the description of existing test.

File: src/xmlstr2xmlnode.js [hunk 3]
```diff
@@ -208,17 +208,17 @@ const getTraversalObj = function(xmlData, options) {
...
-        const tagExp = xmlData.substr(i,closeIndex);
-        if(tagExp.indexOf("[")){
+      } else if( xmlData.substr(i + 1, 2) === '!D') {
+        const closeIndex = xmlData.indexOf(">", i)
+        const tagExp = xmlData.substr(i, closeIndex - i + 1);
+        if(tagExp.indexOf("[") >= 0){
           i = xmlData.indexOf("]>", i) + 1;
         }else{
-          i = closeIndex;
+          i = closeIndex + 1;
```
Comment (amitguptagwl) [inline src/xmlstr2xmlnode.js:219]: It seems not required.

File: spec/cdata_spec.js (hunk 1 above)
Comment (Delagen) [reply spec/cdata_spec.js:298]: What expected?

File: spec/cdata_spec.js (hunk 2 above)
Comment (Delagen) [reply spec/cdata_spec.js:309]: Removed

**Comment by amitguptagwl** [comment]
Thanks for your changes and to update the PR title

**Pull request merged**